internal/controllers/topology/cluster/structuredmerge/dryrun.go

```diff
@@ -0,0 +1,301 @@
+// an actual change when running server side apply, and if this change might impact the object spec or not.
+// NOTE: This func checks if:
+// - something previously managed is missing from intent (a field has been deleted from modified)
+// - the value for a field previously managed is changing in the intent (a field has been changed in modified)
+// - the intent contains something not previously managed (a field has been added to modified).
+func dryRunPatch(ctx *hasChangesContext) (hasChanges, hasSpecChanges bool) {
+	// If the func is processing a modified element of type map
+	if modifiedMap, ok := ctx.modified.(map[string]interface{}); ok {
+		// NOTE: ignoring the error in case the element wasn't in original.
+		originalMap, _ := ctx.original.(map[string]interface{})
+
+		// Process mapType/structType = granular, previously not empty.
+		// NOTE: mapType/structType = atomic is managed like scalar values down in the func;
+		// a map is atomic when the corresponding FieldV1 doesn't have info for the nested fields.
+		if len(ctx.fieldsV1) > 0 {
+			// Process all the fields the modified map
+			keys := sets.NewString()
+			hasChanges, hasSpecChanges = false, false
+			for field, fieldValue := range modifiedMap {
+				// Skip apiVersion, kind, metadata.name and metadata.namespace which are required field for a
+				// server side apply intent but not tracked into metadata.managedFields, otherwise they will be
+				// considered as a new field added to modified because not previously managed.
+				if len(ctx.path) == 0 && (field == "apiVersion" || field == "kind") {
+					continue
+				}
+				if len(ctx.path) == 1 && ctx.path[0] == "metadata" && (field == "name" || field == "namespace") {
+					continue
+				}
+
+				keys.Insert(field)
+
+				// If this isn't the root of the object and there are already changes detected, it is possible
+				// to skip processing sibling fields.
+				if len(ctx.path) > 0 && hasChanges {
+					continue
+				}
+
+				// Get the original value.
+				fieldPath := ctx.path.Append(field)
+
+				// Get the managed field for this key.
+				// NOTE: ignoring the conversion error that could happen when modified has a field not previously managed
+				fieldV1, _ := ctx.fieldsV1[fmt.Sprintf("f:%s", field)].(map[string]interface{})
+
+				// Get the original value.
+				fieldOriginalValue := originalMap[field]
+
+				// Check for changes in the field value.
+				fieldHasChanges, fieldHasSpecChanges := dryRunPatch(&hasChangesContext{
+					path:     fieldPath,
+					fieldsV1: fieldV1,
+					modified: fieldValue,
+					original: fieldOriginalValue,
+				})
+				hasChanges = hasChanges || fieldHasChanges
+				hasSpecChanges = hasSpecChanges || fieldHasSpecChanges
+			}
+
+			// Process all the fields the corresponding managed field to identify fields previously managed being
+			// dropped from modified.
+			for fieldV1 := range ctx.fieldsV1 {
+				if fieldV1 == "." {
```

Comment:
Can we document what case that is? (if we know :) )

Reply:
I did not find explanation for this in the SSA documentation, it is just something we observed during tests.

Reply:
Without any confirmation but my thinking is, SSA adds the "." field to a path which is a map/Set, if the map/Set gets created (does not exist in the old object).

xref: https://github.com/kubernetes-sigs/structured-merge-diff/blob/master/fieldpath/serialize.go#L90